### ISSUE TITLE
Use a separate ObjectMapper per technology

### ISSUE THREAD

**Issue by mhalbritter** [spring-projects/spring-boot#33438]
Currently we use a global `ObjectMapper` at least for WebMVC and for the Elasticsearch client (see https://github.com/spring-projects/spring-boot/issues/33426).

We should think about separating these `ObjectMapper`s, because right now if you change for example the date time format for WebMVC, this changes also the serialization format used for ElasticSearch.

There may be other places which share the same object mapper.

**Comment by manofthepeace** [comment]
Since ObjectMapper is quite heavy, if at least there is a known way, per technology, to know how to override the ObjectMapper it could be nice. Like using the default, by default, but if needed have a sort of bean to override that default, like you have proposed for the elasticsearch case;

```java
@Configuration(proxyBeanMethods = false)
class JacksonJsonpMapperConfiguration {
	@Bean
	JacksonJsonpMapper jacksonJsonpMapper() {
		return new JacksonJsonpMapper();
	}
}
```

**Comment by mhalbritter** [comment]
Yes, maybe we come to the conclusion that the drawbacks are not worth it and we stick to one ObjectMapper. In that case we should at least document the ramifications of that.

**Comment by wilkinsona** [comment]
We addressed the problem with Elasticsearch by ensuring that it uses its own, isolated `ObjectMapper`. We also have support for Actuator using an isolated `ObjectMapper`. Given these, I don't think there are any other ramifications of sticking to one ObjectMapper that we need to document. We can reconsider if we learn of any.

**Comment by Exidex** [comment]
@wilkinsona while this may not be huge problem for the Spring itself, the approach of using single global ObjectMapper falls apart very quick once you start to customize the behavior of said ObjectMapper. Because it is global, once you configure it as a solution to some specific problem that you have, things may and will break in the places that you were not aware even existed. The same instance of ObjectMapper may be used for transient cases like http responses/requests, kafka messages listeners/publishers, or persistent like caches or databases. Each requires separate consideration and has separate requirements for the backwards/forwards compatibility. And situation becomes worse when Jackson introduces breaking changes in major version. Expecting that people have unit tests for all the edge cases is not realistic. So please reconsider

**Comment by wilkinsona** [comment]
Thanks for the feedback, but I don't find that to be a convincing argument to reconsider.

If you have a specific problem, I would expect the solution to be applied in a specific place. Changes to the app-wide configuration should be reserved for things that you want to apply everywhere. This is essentially what we did for Elasticsearch. It has very specific requirements for the format of the JSON that it works with so we corrected the auto-configuration to work with a mapper that was specifically configured to meet those needs.

**Comment by Exidex** [comment]
Having one global instance makes it just too easy to just edit global, especially with how easy spring makes that, and requires knowledge to understand that that is not a good way. Having API be designed in a way that will require the decision to be explicit would be beneficial in my opinion, but now at least I have better understanding of the reasoning, thanks.
